homeassistant/components/vicare/fan.py

```diff
@@ -0,0 +1,192 @@
+        device: PyViCareDevice,
+        translation_key: str,
+    ) -> None:
+        """Initialize the fan entity."""
+        super().__init__(device_config, device, translation_key)
+        self._attr_translation_key = translation_key
+
+        with suppress(PyViCareNotSupportedFeatureError):
+            self._attributes["vicare_modes"] = self._api.getAvailableModes()
+        self._attributes["vicare_programs"] = self._api.getAvailablePrograms()
+
+    def update(self) -> None:
+        """Update state of fan."""
+        try:
+            with suppress(PyViCareNotSupportedFeatureError):
+                self._attributes["active_vicare_mode"] = self._api.getActiveMode()
+            with suppress(PyViCareNotSupportedFeatureError):
+                self._attributes["active_vicare_program"] = self._api.getActiveProgram()
+        except RequestConnectionError:
+            _LOGGER.error("Unable to retrieve data from ViCare server")
+        except ValueError:
+            _LOGGER.error("Unable to decode data from ViCare server")
+        except PyViCareRateLimitError as limit_exception:
+            _LOGGER.error("Vicare API rate limit exceeded: %s", limit_exception)
+        except PyViCareInvalidDataError as invalid_data_exception:
+            _LOGGER.error("Invalid data from Vicare server: %s", invalid_data_exception)
+
+    def turn_off(self, **kwargs: Any) -> None:
+        """Turn the entity off."""
+        raise ServiceValidationError(
+            translation_domain=DOMAIN,
+            translation_key="fan_must_be_on",
+        )
+
+    @property
+    def is_on(self) -> bool | None:
+        """Return true if the entity is on."""
+        # Viessmann ventilation unit cannot be turned off
+        return True
+
+    @property
+    def percentage(self) -> int | None:
+        """Return the current speed percentage."""
+
+        # if fan mode is something else than *permanent*, speed is controlled by sensor/schedule
+        if self._attributes["active_vicare_mode"] is not VentilationMode.PERMANENT:
+            _LOGGER.debug(
+                "percentage disabled, active vicare mode '%s' does not match '%s'",
+                self._attributes["active_vicare_mode"],
+                VentilationMode.PERMANENT,
+            )
+            return None
+
+        if self._attributes["active_vicare_program"] not in ORDERED_NAMED_FAN_SPEEDS:
+            _LOGGER.debug(
+                "percentage disabled, active vicare program '%s' is no fan speed program",
+                self._attributes["active_vicare_program"],
+            )
+            return None
+
+        return ordered_list_item_to_percentage(
+            ORDERED_NAMED_FAN_SPEEDS, self._attributes["active_vicare_program"]
+        )
+
+    def set_percentage(self, percentage: int) -> None:
+        """Set the speed of the fan, as a percentage."""
+
+        if percentage == 0:
+            _LOGGER.error("Ventilation device cannot be turned off")
```

Comment:
Is it useful to log an error here? Maybe either fail the service, or set it to the lowest possible fan speed?

Reply:
done, rely on `percentage_to_ordered_list_item`